text/0015-status-update.md

```diff
@@ -0,0 +1,397 @@
+been focused on porting an existing design to Rust, and that this may be
+misguided.
+
+#### Subsetting is special
+
+A major focus of (and motivation for) this work has been hb-subset. I think this
+has influenced our thinking in important ways. I'm not entirely confident in my
+analysis here, so please correct any bad assumptions. With that said:
+
+- My sense is that the design of hb-subset is largely path-dependent on the
+  existing design of HarfBuzz. This is just an intuition. What this means to me,
+  though, is that if we were designing a subsetter from scratch, we might make
+  different choices.
+- Subsetting really feels like a special case. To the best of my understanding
+  it is used almost exclusively(?) by Google Fonts, and at a much larger scale.
+- I think that trying to address the *general* case in a way that also natively
+  supports the subsetting case might be a mistake. In particular I worry that
+  this dramatically constrains our solution space, and will lead to a design
+  that is seriously compromised in the general case.
+- I would prefer to focus on the general case, and then try to figure out a more
+  holistic approach to the subsetting problem. In practical terms, this means a
+  general approach that does more transient allocation for subtables etc, and
+  then a subsetting approach that (for instance) uses a long-running process to
+  handle multiple subsetting requests, and reuses memory and other state between
+  them.
+
+## next steps
+
+### a design for mutability
+
+My main concern right now is figuring out a model for compilation & mutability.
+For compilation, I'm currently imagining generating builder types, but I haven't
+dug into this yet.
+
+Offset resolution and rewriting will also be an important consideration. A
+compiler author should be able to use these tools to do things like deduplicate
+identical subtables, or to reorder tables as needed, which means we do need a
+mechanism for rewriting offsets as needed. There are various options here; for
+instance we could track offsets by ids, and then keep track of the location of
+the offsets as we write tables into a shared buffer, rewriting them later as
+needed. This will need more thought, experimentation, and discussion.
+
+### evaluating current performance
+
+It may also be useful to get some concrete numbers on the efficiency of the
+current work-to-date. I have spent a bit of time looking at disassembly and am
+generally confident that we should be able to achieve good results, but I
+haven't *demonstrated* this.
+
+I'm not sure this is strictly necessary at this point, but it might be helpful;
+if performance is actually terrible for some unforeseen reason, we might decide
+that our approach needs to be fundamentally rethought.
+
+I've also resisted this because I'm not sure of exactly what we should measure.
+In any case, it is something we will want to do at some point.
```

Comment:
I think we want to do this next for the specific reason you noted "if performance is actually terrible for some unforeseen reason, we might decide that our approach needs to be fundamentally rethought."

I agree with you that perf is awful is not the expected outcome but we should demonstrate it.

Reply:
With that in mind, this is somewhere where I would appreciate some input. What sort of benchmark(s) would we find satisfying? Are there some existing HarfBuzz benchmarks I could reuse?

Reply:
As we discussed, I'm working on a more modular outline loader/scaler/hinter that we should be able to retarget to this parser. This is likely a better option for more immediate benchmarks. Testing against HarfBuzz would require a full shaper and that's going to be a significant time investment, even relying on the code in swash.